### PR TITLE
Transaction Logging in Optimistic Transactions

### PR DESCRIPTION
**Pull request by Maithem**
Added transaction logging support for optimistic transactions.

### CODE REVIEW

File: test/src/test/java/org/corfudb/runtime/view/ObjectsViewTest.java [hunk 1]
```diff
@@ -79,20 +79,21 @@ public void canAbortNoTransaction()
     @SuppressWarnings("unchecked")
     public void abortedTransactionDoesNotConflict()
             throws Exception {
+        String mapA = "map a";
```
Comment (no2chem) [inline test/src/test/java/org/corfudb/runtime/view/ObjectsViewTest.java]: needs to be final

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java [hunk 2]
```diff
@@ -236,14 +246,20 @@ public long commitTransaction() throws TransactionAbortedException {
...
+
         // Now we obtain a conditional address from the sequencer.
         // This step currently happens all at once, and we get an
         // address of -1L if it is rejected.
         long address = this.builder.runtime.getStreamsView()
                 .acquireAndWrite(
 
                         // a set of stream-IDs that contains the affected streams
-                        writeSet.keySet(),
+                        affectedStreams,
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: I think that the same thing needs to be done to WriteAfterWriteTransactionContext

Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: there is indeed duplicate code right now in both context, which I hope we can organize better in the future.

File: test/src/test/java/org/corfudb/runtime/view/ObjectsViewTest.java (hunk 1 above)
Comment (Maithem) [inline test/src/test/java/org/corfudb/runtime/view/ObjectsViewTest.java]: done

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java (hunk 2 above)
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: done

**Comment by dahliamalkhi** [comment]
Why is this PR marked bug? is it ready to be merged?

**Comment by Maithem** [comment]
@dahliamalkhi Because it used to be implemented, but then it was accidentally removed. Yes it is.

**Comment by Maithem** [comment]
Rebased.

**Pull request merged**